burn-core/src/optim/rmsprop.rs

```diff
@@ -0,0 +1,500 @@
+}
+
+impl RMSPropMomentum {
+    /// transform [grad](Tensor) and [RMSPropMomentumState] to the next step
+    pub fn transform<B: Backend, const D: usize>(
+        &self,
+        grad: Tensor<B, D>,
+        centered_state: CenteredState<B, D>,
+        momentum_state: Option<RMSPropMomentumState<B, D>>,
+    ) -> (
+        Tensor<B, D>,
+        CenteredState<B, D>,
+        RMSPropMomentumState<B, D>,
+    ) {
+        let grad = grad
+            .clone()
+            .div(centered_state.avg.clone().sqrt().add_scalar(self.epsilon));
+
+        if self.momentum > 0. {
+            let buf = match momentum_state {
+                Some(state) => state
+                    .buf
+                    .clone()
+                    .mul_scalar(self.momentum)
+                    .add(grad.clone()),
+                _ => grad.clone(),
+            };
+            (buf.clone(), centered_state, RMSPropMomentumState { buf })
+        } else {
+            match momentum_state {
+                Some(state) => (grad.clone(), centered_state, state),
+                _ => (
+                    grad.clone(),
+                    centered_state,
+                    RMSPropMomentumState {
+                        buf: Tensor::zeros(grad.shape()),
+                    },
+                ),
+            }
+        }
+    }
+}
+
+/// [RMSPropMomentumState](RMSPropMomentumState) is to store and pass optimizer step params.
+#[derive(Record, Clone, new)]
+pub struct RMSPropMomentumState<B: Backend, const D: usize> {
+    buf: Tensor<B, D>,
+}
+
+impl<B: Backend, const D: usize> RMSPropMomentumState<B, D> {
+    fn to_device(mut self, device: &B::Device) -> Self {
+        self.buf = self.buf.to_device(device);
+        self
+    }
+}
+
+impl<B: Backend> SimpleOptimizer<B> for RMSProp<B> {
+    type State<const D: usize> = RMSPropState<B, D>;
+
+    fn step<const D: usize>(
+        &self,
+        lr: LearningRate,
+        mut tensor: Tensor<B, D>,
+        mut grad: Tensor<B, D>,
+        state: Option<Self::State<D>>,
+    ) -> (Tensor<B, D>, Option<Self::State<D>>) {
+        // fetch state for params
+        let mut state_square_avg = None;
+        let mut state_centered = None;
+        let mut state_momentum = None;
+        if let Some(state) = state {
+            state_square_avg = Some(state.square_avg);
+            state_centered = Some(state.centered);
+            state_momentum = Some(state.momentum);
+        }
+
+        // weight_decay transform
+        if let Some(weight_decay) = &self.weight_decay {
+            let (grad_out, tensor_out) = weight_decay.transform_temp_fix(grad, tensor);
+            grad = grad_out;
+            tensor = tensor_out;
+        }
+
+        // square_avg transform
+        let (mut grad, state_square_avg) =
+            SquareAvgState::transform(self.alpha, grad, state_square_avg);
+
+        // centered trnsform
+        let (grad_out, state_square_avg, state_centered) = CenteredState::transform(
```

Comment:
I think we should extract the `if centered` here instead of inside the `CenteredState::transform` function, since it's not its job to also calculate the other transform that doesn't have a state.

Reply:
Can see the comment below. I would prefer keeping the `step` more simple and straightforward, because the `avg` <- `square_avg` can be placed within `centered transform`, which resembles the form of PyTorch.  (But they optimize it with a further step with square root value)